ushahidi/platform-client
# 1677 caption field optional

issue https://github.com/ushahidi/platform/issues/1677 
### This PR makes it possible to not have the caption field in the post creation form. 
![screen shot 2017-09-01 at 6 18 11 pm](https://user-images.githubusercontent.com/2434401/29988120-fdbd7768-8f41-11e7-888a-9758757fede5.png)

### This pull request makes the following changes:
- Added a 'disable caption' option in the UI so that admins can choose to have/not have captions in their media.
- Added captionDisabled field in config array to toggle on/off the input for media caption field
- UI/DATA:  hides captions when the user disables the caption field. (Jess's feedback) . 

### TODO tasks
* [x] more testing instructions should be added.
* [x] go through notes, get help or feedback => I will wait on Robbie's feedback before I move any further 
* [x] UI: add a note for users telling them that they can re-enable the captions at any time by turning them back on. (Jess's feedback) . 
* [x] UI/DATA:  hide captions when the user disables the caption field. (Jess's feedback) . 


### Notes
* Should I rename the captionDisabled field to "captionEnabled" ? It always feels a bit weird to have a name like `fieldnameDisabled`, but I don't know what the best way to go would be here. Is captionEnabled even an improvement as a name for this config option?  
     * * IMPORTANT:* If we go for captionEnabled we will have to migrate and by default set it up as true, otherwise it would break existing deployments. That should be a very simple migration but I'm noting it here so I don't forget. 

* About the `FIXME` note inside the file attribute-editor.directive.js , line 15: I tried some ideas to solve this from the backend to ensure the data was consistent and not need this line, but the backend seems to be forcing any empty values for that field into an empty array because of the *json flag in the getDefinition and the 'backendoptionJsonAlwaysArray' option on the  `transformJson` method of the data transformer. Do we have some way to make sure the backend transformer creates an object literal instead of an array in that field? 

### Testing checklist:

- [x] When I add an image field to a survey, I have the option to enable its **caption** . 
    -    Add an image field 
    -    Use the toggle UI to enable the caption for that image field.
    -    Save the field
    -    save the survey 
    -    try to add a post for that survey type.
          -  [x] you will see an image field with a caption field. 
- [x] When I **edit** a survey's image field and disable its caption, I can no longer add a caption for that field when I add new posts.
    -   Edit a survey, and edit a pre-existing image field.
    -   Use the toggle UI to disable the image field caption
    -    Save the field
    -    save the survey 
    -  [x]   try to add a post for that survey type.
          - [x]  you will see an image field with no caption. 

- [x] When I disable an image's caption and there's already data for it, I stop seeing the caption in pre-existing posts.
    -    Edit a survey, and edit a pre-existing image field.  That field should have data already preloaded for the images and the images' captions.
    -    Use the toggle UI to **disable** the caption field.
    -    Save the field.
    -    Save the survey 
    -    Select a post (from the same survey) that has an image _with a caption_, on the field you just disabled the caption for.
          - [x] you will see an image but not a caption

- [x] After disabling an image's caption for image fields with pre-existing data, I can decide to re-enable the caption field and see the caption data again.
    -    Edit a survey, and edit a pre-existing image field where the caption was disabled. That field should have data already preloaded for the images and the images' captions.
    -    Use the toggle UI to **enable** the caption field.
    -    Save the field.
    -    Save the survey 
    -    Select a post (from the same survey) that has an image _with a caption_, on the field you just disabled the caption for.
          - [x] you will see an image AND the caption that the image had before it was disabled.
          
Fixes ushahidi/platform#1677 .

Ping @ushahidi/platform
